### PR TITLE
Fix guessing types for filters

### PR DESCRIPTION
**Pull request by franmomu**
## Subject

I am targeting this branch, because these changes are BC.

## Changelog

```markdown
### Changed
- Use FQCN for filter types
```

I've updated the list of field types (removing some that not exist) from [here](https://docs.mongodb.com/manual/reference/bson-types/)

### CODE REVIEW

File: src/Guesser/TypeGuesser.php [hunk 1]
```diff
@@ -61,8 +59,8 @@ public function guessType($class, $property, ModelManagerInterface $modelManager
             case 'float':
                 return new TypeGuess('number', [], Guess::MEDIUM_CONFIDENCE);
             case 'integer':
-            case 'bigint':
```
Comment (core23) [inline src/Guesser/TypeGuesser.php]: Can we remove this? IMHO we should keep support for this and throw a deprecation warning instead

**Comment by franmomu** [comment]
> The nonexisting types were wrong or they were defined in an old version of the extension?
> I wonder if could we add some check to avoid the same situation in the future.

I think they may were added copying them from the SonataDoctrineORMAdminBundle, what I did now is using the [defined constants](https://github.com/doctrine/mongodb-odm/blob/master/lib/Doctrine/ODM/MongoDB/Types/Type.php#L26).

File: src/Guesser/TypeGuesser.php (hunk 1 above)
Comment (franmomu) [reply src/Guesser/TypeGuesser.php]: Ok, I'll add them just in case, but apparently [those type don't exist](https://github.com/doctrine/mongodb-odm/blob/master/lib/Doctrine/ODM/MongoDB/Types/Type.php#L26).

Comment (OskarStark) [reply src/Guesser/TypeGuesser.php]: Thanks, I agree with @core23 here, as this could lead to a weird behavior + it’s an easy way to deprecate the options 👍🏻

File: src/Guesser/FilterTypeGuesser.php [hunk 2]
```diff
@@ -73,36 +82,46 @@ public function guessType($class, $property, ModelManagerInterface $modelManager
...
         switch ($metadata->getTypeOfField($propertyName)) {
-            case 'boolean':
+            case Type::BOOL:
+            case Type::BOOLEAN:
                 $options['field_type'] = BooleanType::class;
                 $options['field_options'] = [];
 
-                return new TypeGuess('doctrine_mongo_boolean', $options, Guess::HIGH_CONFIDENCE);
-//            case 'datetime':
-//            case 'vardatetime':
```
Comment (core23) [inline src/Guesser/FilterTypeGuesser.php]: Why did you remove this? This is a BC break

Comment (franmomu) [reply src/Guesser/FilterTypeGuesser.php]: If you mean 
```php
return new TypeGuess('doctrine_mongo_boolean', $options, Guess::HIGH_CONFIDENCE);
```
has been replaced by 
```php
return new TypeGuess(BooleanFilter::class, $options, Guess::HIGH_CONFIDENCE);
```
The rest of this code was commented out and I've uncommented it using the constant types where the type existed.

Comment (core23) [reply src/Guesser/FilterTypeGuesser.php]: Oh, you're right :smile:

File: src/Guesser/FilterTypeGuesser.php [hunk 3]
```diff
@@ -73,36 +82,46 @@ public function guessType($class, $property, ModelManagerInterface $modelManager
...
-//            case 'datetimetz':
-//                return new TypeGuess('doctrine_orm_datetime', $options, Guess::HIGH_CONFIDENCE);
-//            case 'date':
-//                return new TypeGuess('doctrine_orm_date', $options, Guess::HIGH_CONFIDENCE);
+                return new TypeGuess(BooleanFilter::class, $options, Guess::HIGH_CONFIDENCE);
+            /* @deprecated This type was deprecated since version 3.x, to be removed in 4.0 */
```
Comment (phansys) [inline src/Guesser/FilterTypeGuesser.php]: IMO, these messages should trigger real deprecation errors in order to let the user do the required changes.

File: src/Guesser/TypeGuesser.php [hunk 4]
```diff
@@ -44,31 +45,88 @@ public function guessType($class, $property, ModelManagerInterface $modelManager
...
+                    E_USER_DEPRECATED
+                );
+
+                return new TypeGuess('array', [], Guess::HIGH_CONFIDENCE);
+            case Type::BOOL:
+            case Type::BOOLEAN:
                 return new TypeGuess('boolean', [], Guess::HIGH_CONFIDENCE);
             case 'datetime':
+                @trigger_error(
+                    'The array datetime is deprecated since sonata-project/doctrine-mongodb-admin-bundle 3.x, to be removed in 4.0.'.
```
Comment (phansys) [inline src/Guesser/TypeGuesser.php]:
```suggestion
                    'The datetime type is deprecated since sonata-project/doctrine-mongodb-admin-bundle 3.x, to be removed in 4.0.'.
```

File: src/Guesser/TypeGuesser.php [hunk 5]
```diff
@@ -44,31 +45,88 @@ public function guessType($class, $property, ModelManagerInterface $modelManager
...
+                    E_USER_DEPRECATED
+                );
+
+                return new TypeGuess('datetime', [], Guess::HIGH_CONFIDENCE);
             case 'vardatetime':
+                @trigger_error(
+                    'The array vardatetime is deprecated since sonata-project/doctrine-mongodb-admin-bundle 3.x, to be removed in 4.0.'.
```
Comment (phansys) [inline src/Guesser/TypeGuesser.php]:
```suggestion
                    'The vardatetime type is deprecated since sonata-project/doctrine-mongodb-admin-bundle 3.x, to be removed in 4.0.'.
```

File: src/Guesser/TypeGuesser.php [hunk 6]
```diff
@@ -44,31 +45,88 @@ public function guessType($class, $property, ModelManagerInterface $modelManager
...
+                    E_USER_DEPRECATED
+                );
+
+                return new TypeGuess('datetime', [], Guess::HIGH_CONFIDENCE);
             case 'datetimetz':
-            case 'timestamp':
+                @trigger_error(
+                    'The datetimetz datetime is deprecated since sonata-project/doctrine-mongodb-admin-bundle 3.x, to be removed in 4.0.'.
```
Comment (phansys) [inline src/Guesser/TypeGuesser.php]:
```suggestion
                    'The datetimetz type is deprecated since sonata-project/doctrine-mongodb-admin-bundle 3.x, to be removed in 4.0.'.
```

File: src/Guesser/TypeGuesser.php [hunk 7]
```diff
@@ -44,31 +45,88 @@ public function guessType($class, $property, ModelManagerInterface $modelManager
...
+                return new TypeGuess('datetime', [], Guess::HIGH_CONFIDENCE);
+            case Type::DATE:
+            // NEXT_MAJOR: Use only the constant when dropping support for doctrine/mongodb-odm 1.3.
+            // case Type::DATE_IMMUTABLE:
+            case 'date_immutable':
                 return new TypeGuess('date', [], Guess::HIGH_CONFIDENCE);
             case 'decimal':
-            case 'float':
+                @trigger_error(
+                    'The decimal datetime is deprecated since sonata-project/doctrine-mongodb-admin-bundle 3.x, to be removed in 4.0.'.
```
Comment (phansys) [inline src/Guesser/TypeGuesser.php]:
```suggestion
                    'The decimal type is deprecated since sonata-project/doctrine-mongodb-admin-bundle 3.x, to be removed in 4.0.'.
```

File: src/Guesser/TypeGuesser.php [hunk 8]
```diff
@@ -44,31 +45,88 @@ public function guessType($class, $property, ModelManagerInterface $modelManager
...
                 return new TypeGuess('number', [], Guess::MEDIUM_CONFIDENCE);
-            case 'integer':
+            case Type::FLOAT:
+                return new TypeGuess('number', [], Guess::MEDIUM_CONFIDENCE);
+            case Type::INTEGER:
+            case Type::INT:
+                return new TypeGuess('integer', [], Guess::MEDIUM_CONFIDENCE);
             case 'bigint':
+                @trigger_error(
+                    'The bigint datetime is deprecated since sonata-project/doctrine-mongodb-admin-bundle 3.x, to be removed in 4.0.'.
```
Comment (phansys) [inline src/Guesser/TypeGuesser.php]:
```suggestion
                    'The bigint type is deprecated since sonata-project/doctrine-mongodb-admin-bundle 3.x, to be removed in 4.0.'.
```

File: src/Guesser/TypeGuesser.php [hunk 9]
```diff
@@ -44,31 +45,88 @@ public function guessType($class, $property, ModelManagerInterface $modelManager
...
+                    E_USER_DEPRECATED
+                );
+
+                return new TypeGuess('integer', [], Guess::MEDIUM_CONFIDENCE);
             case 'smallint':
+                @trigger_error(
+                    'The smallint datetime is deprecated since sonata-project/doctrine-mongodb-admin-bundle 3.x, to be removed in 4.0.'.
```
Comment (phansys) [inline src/Guesser/TypeGuesser.php]:
```suggestion
                    'The smallint type is deprecated since sonata-project/doctrine-mongodb-admin-bundle 3.x, to be removed in 4.0.'.
```

File: src/Guesser/TypeGuesser.php [hunk 10]
```diff
@@ -44,31 +45,88 @@ public function guessType($class, $property, ModelManagerInterface $modelManager
...
+                    E_USER_DEPRECATED
+                );
+
                 return new TypeGuess('integer', [], Guess::MEDIUM_CONFIDENCE);
-            case 'string':
+            case Type::STRING:
                 return new TypeGuess('text', [], Guess::MEDIUM_CONFIDENCE);
             case 'text':
+                @trigger_error(
+                    'The text datetime is deprecated since sonata-project/doctrine-mongodb-admin-bundle 3.x, to be removed in 4.0.'.
```
Comment (phansys) [inline src/Guesser/TypeGuesser.php]:
```suggestion
                    'The text type is deprecated since sonata-project/doctrine-mongodb-admin-bundle 3.x, to be removed in 4.0.'.
```

File: src/Guesser/TypeGuesser.php [hunk 11]
```diff
@@ -44,31 +45,88 @@ public function guessType($class, $property, ModelManagerInterface $modelManager
...
+                    E_USER_DEPRECATED
+                );
+
                 return new TypeGuess('textarea', [], Guess::MEDIUM_CONFIDENCE);
             case 'time':
+                @trigger_error(
+                    'The decimal datetime is deprecated since sonata-project/doctrine-mongodb-admin-bundle 3.x, to be removed in 4.0.'.
```
Comment (phansys) [inline src/Guesser/TypeGuesser.php]:
```suggestion
                    'The time type is deprecated since sonata-project/doctrine-mongodb-admin-bundle 3.x, to be removed in 4.0.'.
```

File: src/Guesser/TypeGuesser.php (hunk 4 above)
Comment (franmomu) [reply src/Guesser/TypeGuesser.php]: 🤦‍♂

**Comment by core23** [comment]
Thanks @franmomu
